server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt

```diff
@@ -109,133 +117,170 @@ object Chapter {
-                    this[ChapterTable.url] = it.url
-                    this[ChapterTable.name] = it.name
-                    this[ChapterTable.date_upload] = it.uploadDate
-                    this[ChapterTable.chapter_number] = it.chapterNumber
-                    this[ChapterTable.scanlator] = it.scanlator
-                    this[ChapterTable.sourceOrder] = it.index
-                    this[ChapterTable.fetchedAt] = it.fetchedAt
-                    this[ChapterTable.manga] = it.mangaId
-                    this[ChapterTable.realUrl] = it.realUrl
+                    if (chapterEntry == null) {
+                        val newChapterData =
+                            if (chapterData.uploadDate == 0L) {
+                                val altDateUpload = if (maxSeenUploadDate == 0L) now else maxSeenUploadDate
+                                chapterData.copy(uploadDate = altDateUpload)
+                            } else {
+                                maxSeenUploadDate = max(maxSeenUploadDate, chapterData.uploadDate)
+                                chapterData
+                            }
+                        maxSeenUploadDate = max(maxSeenUploadDate, chapterData.uploadDate)
+                        chaptersToInsert.add(newChapterData)
+                    } else {
+                        chaptersToUpdate.add(chapterData)
+                    }
                 }
-            }
 
-            if (chaptersToUpdate.isNotEmpty()) {
-                BatchUpdateStatement(ChapterTable).apply {
-                    chaptersToUpdate.forEach {
-                        addBatch(EntityID(it.id, ChapterTable))
-                        this[ChapterTable.name] = it.name
-                        this[ChapterTable.date_upload] = it.uploadDate
-                        this[ChapterTable.chapter_number] = it.chapterNumber
-                        this[ChapterTable.scanlator] = it.scanlator
-                        this[ChapterTable.sourceOrder] = it.index
-                        this[ChapterTable.realUrl] = it.realUrl
+                val deletedChapterNumbers = TreeSet<Float>()
+                val deletedReadChapterNumbers = TreeSet<Float>()
+                val deletedBookmarkedChapterNumbers = TreeSet<Float>()
+                val deletedChapterNumberDateFetchMap = mutableMapOf<Float, kotlin.Long>()
+
+                // clear any orphaned/duplicate chapters that are in the db but not in `chapterList`
+                val dbChapterCount = chaptersInDb.count()
+                if (dbChapterCount > chapterList.size) { // we got some clean up due
+                    val chapterUrls = chapterList.map { it.url }.toSet()
+
+                    val chaptersIdsToDelete =
+                        chaptersInDb.mapNotNull { dbChapter ->
+                            if (!chapterUrls.contains(dbChapter.url)) {
+                                if (dbChapter.read) deletedReadChapterNumbers.add(dbChapter.chapterNumber)
```

Comment:
are downloads not possible to be preserved?

Reply:
Added this, I am a bit hesitant due to chapters relying on chapter names, and if they are renamed it won't work. But since we don't handling anything like that yet I added it as is.